dev-docs/design/ssr-react.md

```diff
@@ -0,0 +1,96 @@
+
+## Scenarios to Handle
+
+<!-- prettier-ignore-start -->
+| | With vanilla React | What we want |
+| --- | --- | --- |
+| Bare custom elements | <p>Here we simply rely on React's built in SSR which will render the custom element tag, setting all props as attributes. There's no content until JS with component definition is loaded in the client and the custom element is upgraded.</p><p>There is potential for FOUC when the component has light DOM children as these won't have a slot to project to and show up unstyled in the initial document before upgrade, and content shifts as upgraded components fill up spaces.</p> | <p>To mirror the client-side React behavior, all props will be set as attributes on the element as it is server rendered with `@lit-labs/ssr`.<p><p>This will work fine for simple elements whose properties are all serializable and the attribute name matches the property name. More complex components may fail to render.</p> |
+| `@lit-labs/react` wrapped components | <p>The experience here is mostly the same as bare custom elements. No deep server rendering happens. Only the host element tag is emitted. The wrapper prevents props that exist on the component as properties from being emitted as attributes. Component renders client side when definition is loaded.</p><p>The situation with FOUC and content shift exists just like bare custom elements.</p> | <p>The `createComponent` wrapper already separates props to those that should be set as properties on the element and those that can be passed to React to be set as attributes. In the server, the wrapper should specifically pass those that need to be set as properties on the element in a special bag so the element renderer can set as properties before rendering the template shadowroot.</p><p>On the client, Lit element hydration should be deferred until the wrapper has chance to set properties on the element.</p> |
+<!-- prettier-ignore-end -->
+
+<!--
+
+## Scenarios to Handle
+
+### Bare custom elements
+
+#### With vanilla React
+
+Here we simply rely on React's built in SSR which will render the custom element tag, setting all props as attributes. There's no content until JS with component definition is loaded in the client and the custom element is upgraded.
+
+There is potential for FOUC when the component has light DOM children as these won't have a slot to project to and show up unstyled in the initial document before upgrade, and content shifts as upgraded components fill up spaces.
+
+#### What we want
+
+To mirror the client-side React behavior, all props will be set as attributes on the element as it is server rendered with `@lit-labs/ssr`.
+
+This will work fine for simple elements whose properties are all serializable and the attribute name matches the property name. More complex components may fail to render.
```

Comment:
What does fail to render mean? Is it behavior worse than the vanilla React behavior? Or the same?

Reply:
Worth checking in a test to show what happens in these cases?

Reply:
It would largely depend on the component's `render()` and the prop it takes. If it's expecting an object property to always exist (without guards or optional chains), it could throw. These cases would've thrown during client rendering with React as well.

Even with guards, this'd mean the rendered result would not be the same as what it would be if props were set as properties. Again, still the same behavior as React client rendering.

I see the argument for a test showcasing what happens, but I don't think adding tests expecting failures for these is as helpful for maintenance.